### PR TITLE
lib/debug.nix overhaul

### PR DESCRIPTION
**Pull request by Profpatsch**
The `debug` module is very useful for day-to-day nix programming, but hasn’t seen a lot of care and aquired quite a lot of cruft.

Main changes:
* add `*Val*Fn` group of functions
* remove deprecated `strict` function
* deprecate the `showVal` group of functions because `traceSeqN` should be strictly more general
* deprecate the `Xml`-group of functions, because they are mostly trivial and/or overly specific
* deprecate `traceValIfNot` and `addErrorContextToAttrs`
* document the module and tracing functions

More detailed reasoning is included in each commit description.

### CODE REVIEW

**Comment by makefu** [comment]
I think it would be nice to have a consistent write-up (or at least a reference entry) on how to use these functions, not only in the comment field above the code but in the manual itself. This is related to https://github.com/NixOS/nix/issues/701 and 
https://github.com/NixOS/nix/pull/702

**Comment by Profpatsch** [comment]
@makefu I can mention the debug module in the manual (similar to how generators is mentioned), but would rather not duplicate the docstrings (because that *will* go out of sync sooner rather than later).

If there’s a debugging tutorial, it should most likely go into a new chapter of the [nix pills](https://github.com/NixOS/nix-pills/tree/master/pills), which have been designated as our official tutorial as far as I can see.

**Comment by makefu** [comment]
@Profpatsch For newcomers that kind of "documentation" is far from optimal. When you start with nix and nixpkgs it is not very obvious to look for solutions inside the implementation source code.

IMHO the nixpkgs `libs` need a reference similar to what we have for the modules and packages, however to solve this is out of scope for this PR which looks fine.

File: lib/default.nix [hunk 1]
```diff
@@ -117,7 +117,7 @@ let
       traceXMLVal traceXMLValMarked traceSeq traceSeqN traceValSeq
       traceValSeqFn traceValSeqN traceValSeqNFn traceShowVal
       traceShowValMarked showVal traceCall traceCall2 traceCall3
-      traceValIfNot runTests testAllTrue strict traceCallXml
+      traceValIfNot runTests testAllTrue traceCallXml
```
Comment (grahamc) [inline lib/default.nix]: Needs release notes.

File: lib/debug.nix [hunk 2]
```diff
@@ -56,19 +56,25 @@ rec {
   traceShowVal = x: trace (showVal x) x;
   traceShowValMarked = str: x: trace (str + showVal x) x;
   attrNamesToStr = a: lib.concatStringsSep "; " (map (x: "${x}=") (attrNames a));
-  showVal = x:
```
Comment (grahamc) [inline lib/debug.nix]: It isn't clear to me that traceSeqN is an improvement over showVal for this use case. Can you provide some examples of the two's output?

File: lib/debug.nix [hunk 3]
```diff
@@ -76,6 +78,15 @@ rec {
...
 
+  traceXMLVal = x:
+    trace ( "Warning: `traceXMLVal` is deprecated "
+          + "and will be removed in the next release." )
+    (trace (builtins.toXML x) x);
+  traceXMLValMarked = str: x:
+    trace ( "Warning: `traceXMLValMarked` is deprecated "
+          + "and will be removed in the next release." )
+    (trace (str + builtins.toXML x) x);
+
```
Comment (grahamc) [inline lib/debug.nix]: These need release notes.

File: nixos/modules/misc/nixpkgs.nix [hunk 4]
```diff
@@ -33,7 +33,7 @@ let
   configType = mkOptionType {
     name = "nixpkgs-config";
     description = "nixpkgs config";
-    check = traceValIfNot isConfig;
+    check = x: if isConfig x then true else lib.traceSeqN 1 x false;
```
Comment (grahamc) [inline nixos/modules/misc/nixpkgs.nix]: I prefer the named one, actually.

**Comment by Profpatsch** [comment]
@makefu Added an initial section to the manual in the last commit. It is more of a stub for now (but at least users now find something when they search for debug (in the right manual)).

**Comment by grahamc** [comment]
I think all the functions which have been deprecated need:

1. release notes
2. their deprecation warnings to include what to replace them with

File: lib/debug.nix (hunk 2 above)
Comment (Profpatsch) [reply lib/debug.nix]: `showVal` displays the value with one layer of depth. `traceValSeqN 1` does the same, `traceValSeqN 2` does the same for a depth of two layers and so forth.

To be fair, the display is different, but I don’t think there is any information in `showVal` that cannot be seen in the output of `traceValSeqN`.

File: lib/default.nix (hunk 1 above)
Comment (Profpatsch) [reply lib/default.nix]: done

File: lib/debug.nix (hunk 3 above)
Comment (Profpatsch) [reply lib/debug.nix]: done

File: nixos/modules/misc/nixpkgs.nix (hunk 4 above)
Comment (Profpatsch) [reply nixos/modules/misc/nixpkgs.nix]: It’s locally `let`-named in the rebased commit.

**Comment by Profpatsch** [comment]
Added release notes & replacement instructions in the last commit.

File: doc/functions.xml [hunk 5]
```diff
@@ -294,6 +294,22 @@ merge:"diff3"
 
 </section>
 
+<section xml:id="sec-debug">
+  <title>Debugging Nix Expressions</title>
+
+  <para>Nix is a unityped, dynamic language, this means every value can
```
Comment (Mic92) [inline doc/functions.xml]: unityped or untyped?

Comment (shlevy) [reply doc/functions.xml]: Neither of those are true :grin:

Comment (Profpatsch) [reply doc/functions.xml]: @shlevy Hm, why is nix not unityped?

Comment (shlevy) [reply doc/functions.xml]: `[]` and `{}` are different types. The fact that those are checked at eval time (which, FWIW, can be seen as compile time with respect to the underlying `.drv` outputs...) doesn't make them not typed!

Comment (Profpatsch) [reply doc/functions.xml]: Ah, well, I see eval time more as execution time.

**Comment by Profpatsch** [comment]
I rebased it to remove the documentation merge conflicts.

**Comment by Profpatsch** [comment]
Merging because no further feedback.
We can roll back selectively if something breaks (although nothing should, since these are debug functions).

**Pull request merged**